wavesrv/pkg/remote/openai/openai.go

```diff
@@ -68,6 +74,96 @@ func RunCompletion(ctx context.Context, opts *sstore.OpenAIOptsType, prompt []ss
 	return marshalResponse(apiResp), nil
 }
 
+func RunCloudCompletion(ctx context.Context, opts *sstore.OpenAIOptsType, prompt []sstore.OpenAIPromptMessageType) ([]*packet.OpenAIPacketType, error) {
+	if opts == nil {
+		return nil, fmt.Errorf("no openai opts found")
+	}
+	if opts.Model == "" {
+		return nil, fmt.Errorf("no openai model specified")
+	}
+	cloudCompletionRequestConfig := sstore.OpenAICloudCompletionRequest{
+		Model:      opts.Model,
+		Prompt:     prompt,
+		MaxTokens:  opts.MaxTokens,
+		MaxChoices: opts.MaxChoices,
+	}
+	const cloudTestAddr = "http://127.0.0.1:7999/api/chat-completion"
+	payloadBuf := new(bytes.Buffer)
+	json.NewEncoder(payloadBuf).Encode(cloudCompletionRequestConfig)
+	httpreq, err := http.NewRequest("POST", cloudTestAddr, payloadBuf)
+	if err != nil {
+		return nil, fmt.Errorf("request create err: %v", err)
+	}
+	httpresp, err := http.DefaultClient.Do(httpreq)
+	if err != nil {
+		return nil, fmt.Errorf("request err: %v", err)
+	}
+	defer httpresp.Body.Close()
+	body, err := io.ReadAll(httpresp.Body)
+	var apiResp openaiapi.ChatCompletionResponse
+	type jsonResponse struct {
+		Data openaiapi.ChatCompletionResponse
+	}
+	var httpjsonresp jsonResponse
+	err = json.Unmarshal(body, &httpjsonresp)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding json output %v", err)
+	}
+	apiResp = httpjsonresp.Data
+	return marshalResponse(apiResp), nil
+}
+
+func RunCloudCompletionStream(ctx context.Context, opts *sstore.OpenAIOptsType, prompt []sstore.OpenAIPromptMessageType) (chan *packet.OpenAIPacketType, error) {
+	const AWSLambdaCentralWSAddr = "wss://5lfzlg5crl.execute-api.us-west-2.amazonaws.com/dev/"
+	if opts == nil {
+		return nil, fmt.Errorf("no openai opts found")
+	}
+	if opts.Model == "" {
+		return nil, fmt.Errorf("no openai model specified")
+	}
+	conn, _, err := websocket.DefaultDialer.Dial(AWSLambdaCentralWSAddr, nil)
+	if err != nil {
+		log.Printf("Websocket error: %v", err)
+		return nil, fmt.Errorf("Websocket error: %v", err)
+	}
+	cloudCompletionRequestConfig := sstore.OpenAICloudCompletionRequest{
+		Model:      opts.Model,
+		Prompt:     prompt,
+		MaxTokens:  opts.MaxTokens,
+		MaxChoices: opts.MaxChoices,
+	}
+	configMessageBuf := new(bytes.Buffer)
+	json.NewEncoder(configMessageBuf).Encode(cloudCompletionRequestConfig)
+	err = conn.WriteMessage(websocket.TextMessage, configMessageBuf.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("Websocker write config error: %v", err)
+	}
+	rtn := make(chan *packet.OpenAIPacketType, DefaultStreamChanSize)
+	go func() {
```

Comment:
we should think about how to timeout this conn/request if something goes wrong or something hangs.  either a hard timeout after a certain amount of time, or a rolling timeout, like if we don't get a new packet for X seconds, then we kill the connection and exit the loop.  the "killer" can run in a different go routine (as ReadMessage can hang).

Reply:
addressed in doOpenAIStreamCompletion - cmdrunner.go:1526

I added a timeout to the channel, if we don't get a new message sent from the socket read goroutine in enough time, the function sends a timeout error

what do you think about this?
Some issues with this are if the rehdmessage call hangs, this doesn't kill the goroutine bc it's somewhat difficult to do so, so that process will persist taking up resources. If there's a way to set a timeout to web socket read message, I should do that as well 

Also, it sends a timeout error packet so the client loses all the information it had up until the timeout. I think it might be better to add a packet to the output that looks like [Server timeout] or something, so the client can still see what they got up to that point. That's difficult to do tho, because the openai packets have an index and other stuff. We would have to save the last packet in order to create our own

What do you think? is this okay for this release or should I revisit this?

Reply:
The way to "unstick" a file/connection read in go is to close the connection.  So once you do websocketConnection.Close() (or whatever the gorilla equivalent is) that should cause the readmessage call to fail.  (be careful about setting two error messages in this case though)

Reply:
addressed. My solution for this is to pass the web socket connection as a parameter to the channel read function and close the socket when it is done, so the socket always closes. This shouldn't have any issues with double errors because the reader stops reading when it hits its timeout

The gorilla web socket documentation didn't say whether it was ok to call close on a socket twice (in the best case the channel write goroutine and the channel read goroutine both call close one after the other), but I tested it and didn't have any issues